### PR TITLE
Handle inactive users

### PR DESCRIPTION
**Pull request by DanielRyanSmith**
This change introduces two new schecduled tasks. One which notifies any users who have no documented activity in 6 months, and one which removes inactive users with 9 months of inactivity. Users are notified at the 6-month mark and do not receive further notifications. Admins and site editors are excluded from these actions.

### CODE REVIEW

File: api/accounts_api_test.py [hunk 1]
```diff
@@ -63,8 +63,11 @@ def test_create__normal_valid(self):
 
     new_appuser = (user_models.AppUser.query(
         user_models.AppUser.email == 'new_user@example.com').get())
-    self.assertEqual('new_user@example.com', new_appuser.email)
-    self.assertFalse(new_appuser.is_admin)
+    result_email = new_appuser.email
+    result_is_admin = new_appuser.is_admin
+    new_appuser.key.delete()
+    self.assertEqual('new_user@example.com', result_email)
+    self.assertFalse(result_is_admin)
```
Comment (DanielRyanSmith) [inline api/accounts_api_test.py:70]: This change was added to make sure this user that is created for testing is deleted, as it was affecting unrelated tests written for these changes.

File: templates/inactive_user_email.html [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+</br>
+<p>
+  For the sake of site security, a user with 9 months of inactivity is
+  granted normal user access. If you no longer need special access, no
+  action is required on your part.
+  Please note that you if you need special access again, you may contact us at
+  webstatus-request@google.com to re-establish special access.
+</p>
+</br>
+<p>(this is an automated message generated by the WebStatus team)</p>
```
Comment (jrobbins) [inline templates/inactive_user_email.html:20]: this -> This
and
WebStatus -> ChromeStatus.com

Actually, I think we could do without this line.  The message will have a From: line that says "Chromestatus admin@cr-status.appspotmail.com", and our users are pretty sophisticated, so I think they will know that it is an automated email.

**Review by jrobbins — CHANGES_REQUESTED** [review]

File: internals/inactive_users.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
+import logging
+from datetime import datetime, timedelta
+
+from framework.basehandlers import FlaskHandler
+from internals.user_models import AppUser
+
+class RemoveInactiveUsersHandler(FlaskHandler):
+  DEFAULT_LAST_VISIT = (2022, 8, 1)  # 2022-08-01
```
[the remark this answers is not in the document]
Comment (DanielRyanSmith) [reply internals/inactive_users.py]: A simple change - updated!

File: internals/inactive_users.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+      last_visit = user.last_visit
+      # If the user does not have a last visit, it is assumed the last visit
+      # is roughly the date the last_visit field was added.
+      if last_visit is None:
+        y, m, d = self.DEFAULT_LAST_VISIT
+        last_visit = datetime(y, m, d)
+      if last_visit < inactive_cutoff:
+        removed_users.append(user.email)
+        user.delete()
```
[the remark this answers is not in the document]
Comment (DanielRyanSmith) [reply internals/inactive_users.py:46]: Added logging for each removed user

File: internals/notifier.py [hunk 5]
```diff
@@ -256,6 +256,67 @@ def get_feature_starrers(self, feature_id):
...
+      last_visit = user.last_visit
+      # If the user does not have a last visit, it is assumed the last visit
+      # is roughly the date the last_visit field was added.
+      if last_visit is None:
+        y, m, d = self.DEFAULT_LAST_VISIT
+        last_visit = datetime(y, m, d)
+      # Notify if the user has recently fallen inactive in the last 30 days
+      # (we only notify at the 6 month mark, and take action at 9 months)
+      if (last_visit < inactive_cutoff and
+          inactive_cutoff - last_visit <= timedelta(days=30)):
```
[the remark this answers is not in the document]
Comment (DanielRyanSmith) [reply internals/notifier.py]: I used the 2nd solution as `notified_inactive`, which seemed simpler, and made sure to update the field when the user signs in again.

File: templates/inactive_user_email.html (hunk 2 above)
Comment (DanielRyanSmith) [reply templates/inactive_user_email.html]: I see - I've removed this last line as well as making the other suggestion changes 👍

File: internals/notifier_test.py [hunk 6]
```diff
@@ -469,9 +470,14 @@ def test_get_feature_starrers__some_starrers(self):
...
 
+
+    user_1_email = app_user_1.email
+    user_2_email = app_user_2.email
+    app_user_1.key.delete()
+    app_user_2.key.delete()
     actual = notifier.FeatureStar.get_feature_starrers(feature_1_id)
     self.assertCountEqual(
-        [app_user_1.email, app_user_2.email],
+        [user_1_email, user_2_email],
```
Comment (DanielRyanSmith) [inline internals/notifier_test.py:480]: This is another change to remove users that persisted and caused issues in other tests.

File: internals/notifier.py [hunk 7]
```diff
@@ -256,6 +256,66 @@ def get_feature_starrers(self, feature_id):
...
+    for user in users:
+      # Site admins and editors aren't warned due to inactivity.
+      # Also, users that have been previously notified are not notified again.
+      if user.is_admin or user.is_site_editor or user.notified_inactive:
+        continue
+
+      # If the user does not have a last visit, it is assumed the last visit
+      # is roughly the date the last_visit field was added.
+      last_visit = user.last_visit or self.DEFAULT_LAST_VISIT
+      # Notify if the user has recently fallen inactive in the last 30 days
```
Comment (jrobbins) [inline internals/notifier.py:296]: This comment no longer matches the code.

**Review by jrobbins — APPROVED** [review]
LGTM after considering one nit.

**Pull request merged**